### PR TITLE
Discover resource plugins dynamically from the user's project.

### PR DESCRIPTION
**Pull request by CyrusNajmabadi**
Fixes https://github.com/pulumi/pulumi/issues/3438

### CODE REVIEW

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 1]
```diff
@@ -117,37 +118,232 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
...
+		}
+	}
+
+	return &pulumirpc.GetRequiredPluginsResponse{Plugins: plugins}, nil
+}
+
+func (host *dotnetLanguageHost) DetermineDotnetPackageDirectory(ctx context.Context) (string, error) {
+	logging.V(5).Infof("GetRequiredPlugins: Determining package directory")
+
+	//dotnet nuget locals global-packages --list
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go]:
```suggestion
	// Execute: dotnet nuget locals global-packages --list
```

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 2]
```diff
@@ -117,37 +118,232 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
...
+	dir := strings.TrimSpace(commandOutput[index+len("global-packages:"):])
+	logging.V(5).Infof("GetRequiredPlugins: Package directory: %v", dir)
+
+	return dir, nil
+}
+
+func (host *dotnetLanguageHost) DeterminePulumiPackages(ctx context.Context) ([][]string, error) {
+	logging.V(5).Infof("GetRequiredPlugins: Determining pulumi packages")
+
+	// dotnet list package --include-transitive
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go]:
```suggestion
	// Execute: dotnet list package --include-transitive
```

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 3]
```diff
@@ -117,37 +118,225 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
...
+		packageDir, packageName, packageVersion)
+
+	// Check for a `~/.nuget/packages/package_name/package_version/content/version.txt` file.
+
+	versionFilePath := path.Join(packageDir, strings.ToLower(packageName), packageVersion, "content", "version.txt")
+	logging.V(5).Infof("GetRequiredPlugins: version file path: %v", versionFilePath)
+
+	if _, err := os.Stat(versionFilePath); err != nil {
+		if os.IsNotExist(err) {
+			// Pulumi package doesn't contain a version.txt file.  This is not a resource-package.
```
Comment (CyrusNajmabadi) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go]:
```suggestion
			// Pulumi package doesn't contain a version.txt file.  This is not a resource-plugin.
```

**Comment by mikhailshilkov** [comment]
What's the failure mode that we want here? If some obscure error happens when discovering versions in NuGet packages (say, permissions to access the folder), do we want to
a) swallow and continue to printing instructions to manually install plugins as before
b) blow up and quit
c) make this configurable
?

**Comment by CyrusNajmabadi** [comment]
I'm ok with 'b' for now.  though i'll add an env variable to control this behavior so there's an escape hatch if someone does run into a problem.

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 4]
```diff
@@ -117,37 +118,231 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
...
+}
+
+func (host *dotnetLanguageHost) RunDotnetCommand(ctx context.Context, args []string, logToUser bool) (string, error) {
+	commandStr := strings.Join(args, " ")
+	if logging.V(5) {
+		logging.V(5).Infoln("Language host launching process: ", host.exec, commandStr)
 	}
 
 	// Make a connection to the real engine that we will log messages to.
 	conn, err := grpc.Dial(host.engineAddress, grpc.WithInsecure())
```
Comment (pgavlin) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: We shouldn't need to do this per-command. We should be able to connect to the engine up-front and keep that connection around.

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 5]
```diff
@@ -117,37 +118,231 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
...
+}
+
+func (host *dotnetLanguageHost) DetermineDotnetPackageDirectory(ctx context.Context) (string, error) {
+	logging.V(5).Infof("GetRequiredPlugins: Determining package directory")
+
+	// Run the `dotnet nuget locals global-packages --list` command.  Importantly, do not clutter
+	// the stream with the extra steps we're performing. This is just so we can determine the
+	// required plugins.  And, after the first time we do this, subsequent runs will see that the
+	// plugin is installed locally and not need to do anything.
+	args := []string{"nuget", "locals", "global-packages", "--list"}
```
Comment (pgavlin) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: Is there a chance that the user won't have packages installed globally?

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go [hunk 6]
```diff
@@ -117,37 +118,231 @@ func newLanguageHost(exec, engineAddress, tracing string) pulumirpc.LanguageRunt
...
+	// we can examine each package to determine the corresponding resource-plugin for it.
+
+	plugins := []*pulumirpc.PluginDependency{}
+	packageToVersion := make(map[string]string)
+	for _, parts := range pulumiPackages {
+		packageName := parts[0]
+		packageVersion := parts[1]
+
+		if existingVersion := packageToVersion[packageName]; existingVersion == packageVersion {
+			// only include distinct dependencies.
```
Comment (pgavlin) [inline sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: Do we need to support SxS scenarios?

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go (hunk 4 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: Done.

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go (hunk 5 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: From https://docs.microsoft.com/en-us/nuget/consume-packages/managing-the-global-packages-and-cache-folders:

> The global-packages folder is where NuGet installs any downloaded package. Each package is fully expanded into a subfolder that matches the package identifier and version number. Projects using the PackageReference format **always use packages directly from this folder.**

Emphasis mine.

There are no other places that nuget should be looking from.

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go (hunk 6 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: it is *waaaay* too early to ask that now.  i'm at least making sure that if there is a ref that says "i need htis plugin" that we require the plugin.  I'm sure we'll have all sorts of painful sxs stuff we'll have to work through.

Comment (pgavlin) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: It's definitely not too early to ask that. We don't have to support them now, but we absolutely need to start thinking through whether or not we will. And this is a land mine if we do need to in the fullness of time.

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go (hunk 5 above)
Comment (pgavlin) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: Perfect; thanks!

File: sdk/dotnet/cmd/pulumi-language-dotnet/main.go (hunk 6 above)
Comment (CyrusNajmabadi) [reply sdk/dotnet/cmd/pulumi-language-dotnet/main.go]: let me rephrase.  I think the topic is important (and i'll add to the "future work" list).  I just don't think it's something to figure out in this PR :)
